From 762c69a9bae6e72016072d60a6edf2eb7186ca33 Mon Sep 17 00:00:00 2001
From: Kirill Kapranov <kirill.kapranov@compulab.co.il>
Date: Tue, 3 Nov 2020 22:20:00 +0200
Subject: [PATCH 90/91] fdt: Add export SoM data via FDT

Export via FDT the following data:
	board options
	board revision (major)
	product name
	serial number

Signed-off-by: Kirill Kapranov <kirill.kapranov@compulab.co.il>
---
 board/compulab/common/eeprom.c           | 40 +++++++++++++++++++++++---------
 board/compulab/common/eeprom.h           | 22 ++++++++++++++++++
 board/compulab/plat/imx8mm/board/board.c | 28 ++++++++++++++++++++++
 3 files changed, 79 insertions(+), 11 deletions(-)

diff --git a/board/compulab/common/eeprom.c b/board/compulab/common/eeprom.c
index ebcafac04b..95467149f3 100644
--- a/board/compulab/common/eeprom.c
+++ b/board/compulab/common/eeprom.c
@@ -35,17 +35,6 @@
 # define CONFIG_SYS_I2C_EEPROM_BUS_SB	0
 #endif
 
-#define EEPROM_LAYOUT_VER_OFFSET	44
-#define BOARD_SERIAL_OFFSET		20
-#define BOARD_SERIAL_OFFSET_LEGACY	8
-#define BOARD_REV_OFFSET		0
-#define BOARD_REV_OFFSET_LEGACY		6
-#define BOARD_REV_SIZE			2
-#define PRODUCT_NAME_OFFSET		128
-#define PRODUCT_NAME_SIZE		16
-#define MAC_ADDR_OFFSET			4
-#define MAC_ADDR_OFFSET_LEGACY		0
-
 #define LAYOUT_INVALID	0
 #define LAYOUT_LEGACY	0xff
 
@@ -238,6 +227,30 @@ int cl_eeprom_get_product_name(uchar *buf, uint eeprom_bus __attribute__((unused
 	return err;
 }
 
+static int cl_eeprom_read_options(char *buf, const struct eeprom_path *eeprom)
+{
+	int len = 0;
+	int err;
+	uchar tmp[PRODUCT_OPTION_SIZE];
+
+	err = cl_eeprom_setup(eeprom);
+	if (err) {
+		printf("%s: Error accesing i2c %x@%x\n", __func__, eeprom->bus, eeprom->chip);
+		return snprintf(buf, PRODUCT_NAME_SIZE, "unknown");
+	}
+
+	for(int i = 0; i < PRODUCT_OPTION_NUM; ++i) {
+		err = cl_eeprom_read(PRODUCT_OPTION_OFFSET + PRODUCT_OPTION_SIZE * i, tmp, PRODUCT_OPTION_SIZE);
+		if (!err && tmp[0] != 0xff) // Check if the flash isn't written
+			len += snprintf(buf + len, PRODUCT_OPTION_SIZE, (char*)tmp);
+	}
+	return len;
+}
+int cl_eeprom_read_som_options(char *buf)
+{
+	return cl_eeprom_read_options(buf, &eeprom_som);
+}
+
 static int cl_eeprom_read_product_name(char *buf, const struct eeprom_path *eeprom)
 {
 	int len;
@@ -257,6 +270,11 @@ static int cl_eeprom_read_product_name(char *buf, const struct eeprom_path *eepr
 
 	return len;
 }
+int cl_eeprom_read_som_name(char *buf)
+{
+	return cl_eeprom_read_product_name(buf, &eeprom_som);
+}
+
 void cl_eeprom_get_suite(char *buf)
 {
 	buf += cl_eeprom_read_product_name(buf, &eeprom_som);
diff --git a/board/compulab/common/eeprom.h b/board/compulab/common/eeprom.h
index d366f59288..dd7d82d855 100644
--- a/board/compulab/common/eeprom.h
+++ b/board/compulab/common/eeprom.h
@@ -11,11 +11,28 @@
 #define _EEPROM_
 #include <errno.h>
 
+#define EEPROM_LAYOUT_VER_OFFSET	44
+#define BOARD_SERIAL_OFFSET		20
+#define BOARD_SERIAL_OFFSET_LEGACY	8
+#define BOARD_REV_OFFSET		0
+#define BOARD_REV_OFFSET_LEGACY		6
+#define BOARD_REV_SIZE			2
+#define PRODUCT_NAME_OFFSET		128
+#define PRODUCT_NAME_SIZE		16
+#define PRODUCT_OPTION_OFFSET		144
+#define PRODUCT_OPTION_SIZE		16
+#define PRODUCT_OPTION_NUM		5
+#define MAC_ADDR_OFFSET			4
+#define MAC_ADDR_OFFSET_LEGACY		0
+
 #if (defined(CONFIG_SYS_I2C) || defined(CONFIG_DM_I2C))
 int cl_eeprom_read_mac_addr(uchar *buf, uint eeprom_bus);
 u32 cl_eeprom_get_board_rev(uint eeprom_bus);
 int cl_eeprom_get_product_name(uchar *buf, uint eeprom_bus);
+int cl_eeprom_read_som_name(char *buf);
+int cl_eeprom_read_som_options(char *buf);
 void cl_eeprom_get_suite(char* buf);
+void cpl_get_board_serial(struct tag_serialnr *serialnr);
 #else
 static inline int cl_eeprom_read_mac_addr(uchar *buf, uint eeprom_bus)
 {
@@ -29,6 +46,11 @@ static inline int cl_eeprom_get_product_name(uchar *buf, uint eeprom_bus)
 {
 	return -ENOSYS;
 }
+static inline void cpl_get_board_serial(struct tag_serialnr *serialnr)
+{
+	return 0;
+}
+
 #endif
 
 #endif
diff --git a/board/compulab/plat/imx8mm/board/board.c b/board/compulab/plat/imx8mm/board/board.c
index 0caa4dc1b8..0801844bfe 100644
--- a/board/compulab/plat/imx8mm/board/board.c
+++ b/board/compulab/plat/imx8mm/board/board.c
@@ -26,6 +26,7 @@
 #include <power/pmic.h>
 #include <power/bd71837.h>
 #include <usb.h>
+#include <asm/setup.h>
 #include <asm/mach-imx/boot_mode.h>
 #include <asm/mach-imx/video.h>
 #include "ddr/ddr.h"
@@ -86,6 +87,32 @@ int dram_init(void)
 }
 
 #ifdef CONFIG_OF_BOARD_SETUP
+static void fdt_set_sn(void *blob)
+{
+	u32 rev;
+	char buf[100];
+	int len;
+	union {
+		struct tag_serialnr	s;
+		u64			u;
+	} serialnr;
+
+	len = cl_eeprom_read_som_name(buf);
+	fdt_setprop(blob, 0, "product-name", buf, len);
+
+	cpl_get_board_serial(&serialnr.s);
+	fdt_setprop(blob, 0, "serial-number", buf, sprintf(buf, "%llx", serialnr.u) + 1);
+
+	rev = cl_eeprom_get_board_rev(0);
+	fdt_setprop(blob, 0, "board-revision", buf,
+		sprintf(buf, "%u.%02u", rev/100 , rev%100 ) + 1);
+
+	len = cl_eeprom_read_som_options(buf);
+	fdt_setprop(blob, 0, "board-options", buf, len);
+
+	return;
+}
+
 static int fdt_set_env_addr(void *blob, bd_t *bd)
 {
 	char tmp[32];
@@ -110,6 +137,7 @@ static int fdt_set_env_addr(void *blob, bd_t *bd)
 int ft_board_setup(void *blob, bd_t *bd)
 {
 	fdt_set_env_addr(blob, bd);
+	fdt_set_sn(blob);
 	return 0;
 }
 #endif
-- 
2.11.0

